src/lazyload-image.ts

```diff
@@ -65,9 +99,9 @@ function setLoadedStyle(element: HTMLElement) {
     return element;
 }
 
-export function lazyLoadImage(image: HTMLElement, imagePath: string, defaultImagePath: string, errorImgPath: string, offset: number) {
+export function lazyLoadImage(image: HTMLElement, imagePath: string, defaultImagePath: string, errorImgPath: string, offset: number, useSrcset: boolean) {
     if (defaultImagePath) {
-        setImage(image, defaultImagePath);
+        setImage(image, defaultImagePath, useSrcset, 'defaultImage');
```

Comment:
That's a nice catch.
I've added this, except named `useSources` to `isImgOfPicture`, which to me seems a little more clear.
But since this is your project, I'm OK with naming things the way you want.

This can however be shortened, by using a higher order function, like this:
```TypeScript
function setImageAndSources(setSourcesFn: (image: HTMLImageElement) => void) {
    return (element: HTMLImageElement | HTMLDivElement, imagePath: string, useSrcset: boolean) => {
        if (isImageElement(element) && isChildOfPicture(element)) {
            setSourcesFn(element);
        }
        if (imagePath) {
            setImage(element, imagePath, useSrcset);
        }
    }
}

const setImageAndSourcesToDefault = setImageAndSources(setSourcesToDefault);
const setImageAndSourcesToLazy = setImageAndSources(setSourcesToLazy);
const setImageAndSourcesToError = setImageAndSources(setSourcesToError);
```
Might not be the best naming, since it comes out quite long.
But now all 3 parts under `lazyLoadImage` function can be simplified.
From:
```TypeScript
export function lazyLoadImage(element: HTMLImageElement | HTMLDivElement, imagePath: string, defaultImagePath: string, errorImgPath: string, offset: number, useSrcset: boolean = false) {
    const isImgOfPicture = isImageElement(element) && isChildOfPicture(element);
    if (isImgOfPicture) {
        setSourcesToDefault(element as HTMLImageElement);
    }
    if (defaultImagePath) {
        setImage(element, defaultImagePath, useSrcset);
    }
    if (element.className && element.className.includes('ng-lazyloaded')) {
        element.className = element.className.replace('ng-lazyloaded', '');
    }

    return (scrollObservable: Observable<Event>) => {
        return scrollObservable
            .filter(() => isVisible(element, offset))
            .take(1)
            .mergeMap(() => loadImage(element, imagePath, useSrcset))
            .do(() => {
                if (isImgOfPicture) {
                    setSourcesToLazy(element as HTMLImageElement);
                }
                setImage(element, imagePath, useSrcset);
            })
            .map(() => true)
            .catch(() => {
                if (isImgOfPicture) {
                    setSourcesToError(element as HTMLImageElement);
                }
                if (errorImgPath) {
                    setImage(element, errorImgPath, useSrcset);
                }
                element.className += ' ng-failed-lazyloaded';
                return Observable.of(false);
            })
            .do(() => setLoadedStyle(element));
    };
}
```
To simply:
```Typescript
export function lazyLoadImage(element: HTMLImageElement | HTMLDivElement, imagePath: string, defaultImagePath: string, errorImgPath: string, offset: number, useSrcset: boolean = false) {
    setImageAndSourcesToDefault(element);
    if (element.className && element.className.includes('ng-lazyloaded')) {
        element.className = element.className.replace('ng-lazyloaded', '');
    }

    return (scrollObservable: Observable<Event>) => {
        return scrollObservable
            .filter(() => isVisible(element, offset))
            .take(1)
            .mergeMap(() => loadImage(element, imagePath, useSrcset))
            .do(() => setImageAndSourcesToLazy(element))
            .map(() => true)
            .catch(() => {
                setImageAndSourcesToError(element);
                element.className += ' ng-failed-lazyloaded';
                return Observable.of(false);
            })
            .do(() => setLoadedStyle(element));
    };
}
```

This is not yet committed, as I want to hear your opinion first.

Reply:
I love it! We should definitely create `setImageAndSources`, as you suggested!